### PR TITLE
gh-98253: Break potential reference cycles in external code worsened by typing.py lru_cache

### PR DESCRIPTION
**Pull request by wjakob**
The ``typing`` module internally uses LRU caches to memoize the result of type-related computations. These LRU caches have the potential to introduce difficult-to-find reference leaks spanning multiple extension modules. 

Suppose that a binary extension module ``a`` leaks a reference to a type ``a.A`` with typed function signatures.

The leaked type ``a.A`` will induce a secondary refleak of the ``typing`` LRU caches, which will at this point cause tertiary leaks in entirely unrelated extension modules. In this way, a benign refleak in any extension can cause difficult-to-debug leaks everywhere else.

This commit fixes this by removing the direct reference from ``a.A`` to ``typing`` implementation details.

* Issue: gh-98253

### CODE REVIEW

**Comment by wjakob** [comment]
PS: It would be good to merge this not only in Python 3.12, but also older versions (3.11, 3.10, 3.9, 3.8).

**Comment by AlexWaygood** [comment]
> PS: It would be good to merge this not only in Python 3.12, but also older versions (3.11, 3.10, 3.9, 3.8).

3.8 and 3.9 are now only accepting security-related patches, so this definitely won't be backported that far, I'm afraid.

**Comment by wjakob** [comment]
> 3.8 and 3.9 are now only accepting security-related patches, so this definitely won't be backported that far, I'm afraid.

Fair enough. In that case, I propose to target 3.10, 3.11, and 3.12.

**Comment by JelleZijlstra** [comment]
As mentioned in the issue I'm skeptical that this is something we should fix in typing.py. I'll ask for input from other core devs (after the dust from the 3.11 release settles).

**Comment by gvanrossum** [comment]
I'm skeptical too. I feel that the tool is incorrectly blaming typing.py.

**Comment by wjakob** [comment]
@gvanrossum, did you see the discussion in https://github.com/python/cpython/issues/98253? When I first reported the issue, the actual source of these refleaks was mysterious. Further along the discussion thread, the problem was finally tracked down (see post https://github.com/python/cpython/issues/98253#issuecomment-1285359073)

The problem is that the caching mechanism in `typing.py` can turn virtually *any refleak* in *any typed extension* module into refleaks elsewhere. That makes it really difficult to develop leak-free extensions, because any systematic checking for leaks runs into flukes.

The patch in this PR is simple and it breaks this problematic chain of references. Of course, one could say: let's leave `typing.py` as-is and fix all of the other extension modules. I just don't think it is very practical to do so.

**Comment by wjakob** [comment]
> I'm skeptical too. I feel that the tool is incorrectly blaming typing.py.

If by "the tool", you mean the nanobind-based reproducer, please let me know. It would be straightforward to reproduce the leak with a minimal CPython extension.

I can sit down and make such a producer, but I would prefer to only do so if it's time well spent (in the sense that there is willingness to fix the issue if convincingly demonstrated).

**Comment by Fidget-Spinner** [comment]
Speaking as a typing maintainer,

Theoretically, I would like to say that the onus is on the external code to fix their refleaks. Since in this case as you have pointed out, typing.py doesn't have any refleaks. It just worsens existing ones.

However, I'm not opposed to the change (+0.5), because there aren't _that_ many more lines of code. So the maintenance burden to support something like this isn't large.

Also as you have rightly pointed out, it's nearly impossible to get all extension modules to fix all their refleaks. So I'm satisfied with typing not making them worse.

**Comment by gvanrossum** [comment]
@wjakob Never mind. I trust @Fidget-Spinner here.

**Comment by kumaraditya303** [comment]
`static` global state in extension modules is bad and should be avoided as much. See https://peps.python.org/pep-0489/
They are bad for multiple reasons not to mention they are a problem for sub-interpreter. I worry if we add workarounds for this in stdlib then library authors will never fix their own code and we will get more requests to add workarounds and they will never fix their own code.  Note that just clearing reference isn't enough here, when the GC is involved this can lead to very strange bugs and are a pain to debug.  See a similar `asyncio` bug https://github.com/python/cpython/issues/96232 involving `static` c globals. For this specific case, this seems fine but we need to be cautious because if we add this then the extension authors will never notice that there is a bug in their code as it will probably "just work".

P.S I am not a typing maintainer so I'll defer to @Fidget-Spinner.

**Comment by wjakob** [comment]
There are a few misconceptions in the above message, I'll try to clarify. 

> static global state in extension modules is bad and should be avoided as much. [..] I worry if we add workarounds for this in stdlib then library authors will never fix their own code and we will get more requests to add workarounds so that they don't have to fix their own code.

While this PR does introduce a static state variable, it represents state that is already static (the caches). FWIW there is a `_cleanup` variable just a few lines above that serves the exact same purpose, so it does not seem like new ground is being broken.

>  [..] if we add this then the extension authors will never notice that there is a bug in their code as it will probably "just work".

This is already what is happening. The leaks in question don't cause issues for those respective packages.

The problem is that the mechanics typing module spreads these leaks to other modules, which is where it becomes problematic. As mentioned in the associated issue, expecting all other extensions to be fixed instead is is going to be an impossible game of whack-a-mole.

**Comment by kumaraditya303** [comment]
> While this PR does introduce a static state variable, it represents state that is already static (the caches). FWIW there is a _cleanup variable just a few lines above that serves the exact same purpose, so it does not seem like new ground is being broken.

`_cleanups` has a different purpose, it is used by our test runner to check for reference leaks by clearing caches and it not used outside of that purpose.

https://github.com/python/cpython/blob/de6981680bcf6496e5996a853b2eaa700ed59b2c/Lib/test/libregrtest/utils.py#L211-L212

**Comment by gvanrossum** [comment]
> While this PR does introduce a static state variable

There seems to be a miscommunication here. When Kumar wrote "static global state in extension modules is bad" he meant C code (maybe referring to the bug you found in some extension?). This has nothing to do with static state in a .py module like referenced by this PR. So let's please all calm down.

File: Misc/NEWS.d/next/Library/2022-10-24-11-01-05.gh-issue-98253.HVd5v4.rst
Comment (rhettinger) [inline Misc/NEWS.d/next/Library/2022-10-24-11-01-05.gh-issue-98253.HVd5v4.rst]: This doesn't match the investigation results. Neither typing, nor the lru_cache is broken. Neither has "caused refleaks".  Instead, the results of the investigation was that uncleared caches in typing would amplify existing refleaks in third-party extension modules.

This is more properly described as "harm reduction" rather than a bugfix.

**Comment by wjakob** [comment]
@rhettinger : I added a longer paragraph summarizing the issue and incorporated your feedback.

**Comment by AlexWaygood** [comment]
Please don't force-push to CPython PRs after review/discussion has started. It interacts badly with GitHub's UI, making it harder to see what exactly changed since we last looked at the PR :)

**Comment by wjakob** [comment]
Sorry 😇.  For posterity, I amended the commit to replace the contents of the file `Misc/NEWS.d/next/Library/2022-10-24-11-01-05.gh-issue-98253.HVd5v4.rst`.

File: Lib/typing.py
Comment (JelleZijlstra) [inline Lib/typing.py:368]: This will be marginally slower because instead of a single cell variable lookup, we do a global lookup, a cell variable, and a subscript. Probably not enough to matter.

Comment (Fidget-Spinner) [reply Lib/typing.py:368]: Assuming we don't backport this: this should still be potentially faster in 3.12 than in 3.11.

**Comment by Fidget-Spinner** [comment]
I will accept it as an enhancement to typing soon (ie it wont be backported). Sorry for the wait, I was busy with uni exams.

**Comment by Fidget-Spinner** [comment]
@serhiy-storchaka or @JelleZijlstra does the code look okay to y'all? I plan to merge this soon.

File: Lib/typing.py
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
_caches = {}
```

File: Lib/typing.py
Comment (AlexWaygood) [inline Lib/typing.py:363]: Can we add a short comment here about why we're taking this approach (maybe linking back to the GitHub issue)?

**Review by AlexWaygood — COMMENTED** [review]
I'm happy with this to go in, but I think a short comment wouldn't go amiss :) there's already a lot of code in `typing.py` that's quite subtle

File: Lib/typing.py
Comment (wjakob) [reply Lib/typing.py:363]: Done just now.

File: Lib/typing.py
Comment (wjakob) [reply Lib/typing.py]: Done.

**Comment by wjakob** [comment]
It's quite puzzling -- I pushed two further commits to address your feedback, but they don't show up here. Perhaps a GitHub service issue/outage? See https://github.com/python/cpython/compare/main...wjakob:cpython:typing-leak

**Comment by AlexWaygood** [comment]
> It's quite puzzling -- I pushed two further commits to address your feedback, but they don't show up here. Perhaps a GitHub service issue/outage? See [main...wjakob:cpython:typing-leak](https://github.com/python/cpython/compare/main...wjakob:cpython:typing-leak)

GitHub is quite laggy right now in general, I'm seeing it on `typeshed` and `typing_extensions` as well :)

**Comment by JelleZijlstra** [comment]
Yes, GitHub is apparently having some issues; I saw it on other PRs. I did get emails for the two commits you pushed.

**Comment by wjakob** [comment]
Edit:  now, something is happening.

**Review by AlexWaygood — APPROVED** [review]
I'm happy for this to be merged (but will wait for at least one other approval before merging)

**Comment by JelleZijlstra** [comment]
https://www.githubstatus.com/ reports some degradation.

**Review by Fidget-Spinner — APPROVED** [review]

**Pull request merged**
